---
title: "Case study on Covid-19 B-cells: differential gene expression analysis"
author: "Jeroen Gilis & Laura Perin"
date: "14/11/2023"
output:
    html_document:
      code_download: true    
      theme: cosmo
      toc: true
      toc_float: true
      highlight: tango
      number_sections: true
---

# Load libraries

```{r, message=FALSE, warning=FALSE}
library(SingleCellExperiment)
library(scuttle)
library(ggplot2)
library(patchwork)
library(scater)
library(scran)
library(edgeR)
library(limma)
library(pbapply)
```

# Prepare storing figures

```{r}
here::i_am("2_CovidCase_DGE.Rmd")
out_dir <- here::here("figures")
fs::dir_create(out_dir)
```

# Pseudobulk data

```{r}
sce <- readRDS("./objects/sce_Covid_Bcells.rds")

pb <- aggregateAcrossCells(sce,
                           use.assay.type = "counts",
                           ids = colData(sce)[,c("donor_id","cell_type_curated")])

pb$Status_on_day_collection_summary <- factor(pb$Status_on_day_collection_summary, 
                                              levels = c("Healthy", "Asymptomatic", "Mild",
                                                         "Moderate", "Severe", "Critical"))
pb$development_stage <- factor(pb$development_stage, 
                               levels = c("third decade human stage",
                                          "fourth decade human stage",
                                          "fifth decade human stage",
                                          "sixth decade human stage",
                                          "seventh decade human stage",
                                          "eighth decade human stage",
                                          "ninth decade human stage",
                                          "tenth decade human stage"))

pb$sex <- as.factor(pb$sex)
pb$cell_type_curated <- as.factor(pb$cell_type_curated)
```

# Patient-level filtering

```{r}
pb <- pb[,pb$ncells >= 20] # remove psudobulk samples from less than 20 cells
# retain only these five genrally abundant cell types
pb <- pb[,pb$cell_type_curated %in% c("B cell", "class switched memory B cell",
                                      "immature B cell", "naive B cell",
                                      "unswitched memory B cell")] 
pb <- pb[,pb$Site != "Sanger"] # remove samples for the Sanger sequencing site:
# not that many samples and no healthy samples
colData(pb) <- droplevels(colData(pb))
```

# Global gene-level filtering

```{r}
# initial lenient filtering
keep <- rowSums(assay(pb)) > 5
table(keep)
pb <- pb[keep,]
```

# Covariate selection

## Normalization

```{r}
# received error without prior computation of size factors...
sf <- librarySizeFactors(pb)
pb <- logNormCounts(pb,
                    size.factors=sf)
```

## Feature selection

```{r, warning=FALSE}
dec <- modelGeneVar(pb)
fit <- metadata(dec)
# get 10% most variable genes
hvg <- getTopHVGs(dec, 
                  prop=0.1)
```

## PCA

```{r}
set.seed(1234)
pb <- runPCA(pb, 
             ncomponents=30, 
             subset_row=hvg)
```

## t-SNE
 
```{r}
pb <- runTSNE(pb,
              dimred = 'PCA',
              n_dimred = 10,
              external_neighbors=TRUE)

plotTSNE(pb,
         colour_by = "Site")
```

```{r}
gg_cam_1 <- plotTSNE(pb[,pb$Site == "Cambridge"],
                     colour_by = "cell_type_curated")
gg_cam_2 <- plotTSNE(pb[,pb$Site == "Cambridge"],
                     colour_by = "sex")
gg_cam_3 <- plotTSNE(pb[,pb$Site == "Cambridge"],
                     colour_by = "Status")

gg_Ncl_1 <- plotTSNE(pb[,pb$Site == "Ncl"],
                     colour_by = "cell_type_curated")
gg_Ncl_2 <- plotTSNE(pb[,pb$Site == "Ncl"],
                     colour_by = "sex")
gg_Ncl_3 <- plotTSNE(pb[,pb$Site == "Ncl"],
                     colour_by = "Status")
```

```{r, fig.width=10, fig.height=6}
pdf("./figures/tsne_explore_covariates.pdf",
            width     = 12,
            height    = 8,
            pointsize = 4)
(gg_cam_1 + gg_cam_2 + gg_cam_3) /
  (gg_Ncl_1 + gg_Ncl_2 + gg_Ncl_3)
dev.off()
```

I'd say it's vital to include Site and sex in the analysis. It doesn't
look necessary to additionally include age;

```{r, fig.width=6, fig.height=4}
plotTSNE(pb[,which(pb$Site == "Ncl" & pb$cell_type_curated %in% c("immature B cell", "naive B cell"))],
         colour_by = "development_stage",
         shape_by = "Status",
         point_size=3,
         point_alpha = 1) + ggtitle("development_stage")

plotTSNE(pb[,which(pb$Site == "Ncl" & pb$cell_type_curated %in% c("immature B cell", "naive B cell"))],
         colour_by = "Smoker",
         shape_by = "Status",
         point_size=3,
         point_alpha = 1) + ggtitle("Smoker")

plotTSNE(pb[,which(pb$Site == "Ncl" & pb$cell_type_curated %in% c("immature B cell", "naive B cell"))],
         colour_by = "ncells",
         shape_by = "Status",
         point_size=3,
         point_alpha = 1) + ggtitle("ncells")

plotTSNE(pb[,which(pb$Site == "Ncl" & pb$cell_type_curated %in% c("immature B cell", "naive B cell"))],
         colour_by = "Outcome",
         shape_by = "Status",
         point_size=3,
         point_alpha = 1) + ggtitle("Outcome")

plotTSNE(pb[,which(pb$Site == "Ncl" & pb$cell_type_curated %in% c("immature B cell", "naive B cell"))],
         colour_by = "Resample",
         shape_by = "Status",
         point_size=3,
         point_alpha = 1) + ggtitle("Resample")

plotTSNE(pb[,which(pb$Site == "Ncl" & pb$cell_type_curated %in% c("immature B cell", "naive B cell"))],
         colour_by = "Swab_result",
         shape_by = "Status",
         point_size=3,
         point_alpha = 1) + ggtitle("Swab_result")

plotTSNE(pb[,which(pb$Site == "Ncl" & pb$cell_type_curated %in% c("immature B cell", "naive B cell"))],
         colour_by = "Days_from_onset",
         shape_by = "Status",
         point_size=3,
         point_alpha = 1) + ggtitle("Days_from_onset")
```

# Ct-specific gene filtering

```{r}
# Subset the data per cell type
# Filter each cell type separately with edgeR::filterByExpr
design <- model.matrix(~ Site + sex + Status_on_day_collection_summary, 
                       colData(pb))
colnames(design) <- sub("Status_on_day_collection_summary", "", colnames(design))
colnames(design)[1] <- "Intercept"
label <- pb$cell_type_curated
pb_ct <- vector(mode =  "list", length = length(unique(label)))

for (i in 1:length(unique(label))) {
  
  chosen <- label==sort(unique(label))[i]
  pb_i <- pb[,chosen,drop=FALSE]
  design_i <-  design[chosen,]

  gkeep <- filterByExpr(pb_i, design=design_i)
  pb_ct[[i]] <- pb_i[gkeep,]
}
names(pb_ct) <- sort(unique(label))

saveRDS(pb_ct, "./objects/pb_ct_filt.rds")
```

# Setup contrast matrix

```{r}
design <- model.matrix(~ Site + sex + Status_on_day_collection_summary, 
                         colData(pb_ct$`B cell`))
colnames(design) <- sub("Status_on_day_collection_summary", "", colnames(design))
colnames(design)[1] <- "Intercept"
  
L <- matrix(0, ncol = 5, nrow = ncol(design)) # initialize contrast matrix
rownames(L) <- colnames(design)
colnames(L) <- c("AsymptomaticVsHealthy", "MildVsHealthy", 
                 "ModerateVsHealthy", "SevereVsHealthy",
                 "CriticalVsHealthy")

L["Asymptomatic",1] <- 1
L["Mild",2] <- 1
L["Moderate",3] <- 1
L["Severe",4] <- 1
L["Critical",5] <- 1
L
```

# Differential expression analysis - edgeR-NB

## Perform analysis

```{r}
DGE_res <- pblapply(pb_ct, function(element){
    
    # get cell type name
    ct <- levels(droplevels(element$cell_type_curated))
    
    # create design matrix
    design <- model.matrix(~ Site + sex + Status_on_day_collection_summary, 
                           colData(element))
    colnames(design) <- sub("Status_on_day_collection_summary", "", colnames(design))
    colnames(design)[1] <- "Intercept"
    
    # Differential expression analysis with edgeR-NB
    y <- DGEList(counts=counts(element), samples=colData(element))
    y <- calcNormFactors(y)
    y <- estimateDisp(y,design)
    fit <- glmQLFit(y,design,robust=T)
    
    # Intermediate visualization
    par(mfrow=c(1,2))
    plotBCV(y, main = paste0("BCV plot for ", ct), cex.main=0.7)
    plotQLDisp(fit, main = paste0("QLDisp plot for ", ct), cex.main=0.7)
    
    # Differential testing
    resList_i <- vector(mode = "list", length = ncol(L))
    for(j in 1:ncol(L)){
      test <- glmQLFTest(fit, contrast=L[,j])
      resList_i[[j]] <- test$table
    }
    return(resList_i)
})

saveRDS(DGE_res ,"./objects/DGE_res_edgeR_NB.rds")
```

## Visualize p-values

```{r}
pdf("./figures/pval_hist_DGE_edgeR_NB.pdf",
            width     = 8,
            height    = 8,
            pointsize = 4)
par(mfrow=c(5,5))
for (j in 1:5) {
  for(i in 1:5){
    pat_per_ct <- table(pb_ct[[j]]$Status_on_day_collection_summary)
    hist(DGE_res[[j]][[i]]$PValue, breaks=20, xlab = "p-values",
         main = paste0(levels(droplevels(pb_ct[[j]]$cell_type_curated)), " \n ", 
                       levels(droplevels(pb_ct[[j]]$Status_on_day_collection_summary))[i+1], " - ",
                       pat_per_ct[1], " vs ",pat_per_ct[i+1]))
  }
}
dev.off()
```

# Session info

```{r}
sessionInfo()
```





