---
title: "Case study on Covid-19 B-cells: downstream comparison and visualization"
author: "Jeroen Gilis & Laura Perin"
date: "14/11/2023"
output:
    html_document:
      code_download: true    
      theme: cosmo
      toc: true
      toc_float: true
      highlight: tango
      number_sections: true
---

```{r}
here::i_am("4_CovidCase_downstream.Rmd")
```

# Load libraries

```{r, message=FALSE, warning=FALSE}
library(SingleCellExperiment)
library(scuttle)
library(ggplot2)
library(patchwork)
library(scater)
library(scran)
library(edgeR)
library(limma)
library(pbapply)
library(iCOBRA)
library(harmonicmeanp)
library(stageR)
library(dplyr)
library(ComplexHeatmap)
library(openxlsx)
library(biomaRt)
library(ggrepel)
library(GO.db)
```

# Load helper functions

```{r}
source("./5_CovidCase_HelperFunctions.R")
```

# Load DGE analysis data and results

```{r}
pb_ct <- readRDS("./objects/pb_ct_filt.rds")
DGE_res <- readRDS("./objects/DGE_res_edgeR_NB.rds")
```

# Load DD analysis data and results

```{r}
pb_bin_ct <- readRDS("./objects/pb_bin_ct_filt.rds")
DD_res <- readRDS("./objects/DD_res_edgeR_NB_optim.rds")
```

# Venn diagram

This Venn diagram was not used as a figure in our manuscript, nor in the 
manuscript's supplementary materials.

```{r}
pdf("./figures/Venndiagram_DD_DE.pdf",
            width     = 5,
            height    = 3,
            pointsize = 4)
for (j in seq_along(DD_res)) {
  
  levels_status <- levels(pb_bin_ct$`B cell`$Status_on_day_collection_summary)
  
  par(mfrow=c(2,3))
  for(i in 1:5){
    pval_df <- data.frame(DGE = DGE_res[[j]][[i]]$PValue,
                          DD = DD_res[[j]][[i]]$PValue)
    rownames(pval_df) <- rownames(DGE_res[[j]][[i]])
    pad <- pval_df
    pad$DGE <- p.adjust(pad$DGE, method = "BH")
    pad$DD <- p.adjust(pad$DD, method = "BH")
    cobradata <- COBRAData(pval = pval_df, padj=pad)
    cobraperf <- calculate_performance(cobradata, splv = "none",maxsplit = 4)
    cobraplot <- prepare_data_for_plot(cobraperf, colorscheme = "Dark2", facetted = TRUE)
    
    plot_overlap(cobraplot, main = paste0(names(DD_res)[j], "\n", levels_status[i+1]),
                 mar = rep(2,4))
  }
}
dev.off()
```

# Stagewise testing

```{r, warning=FALSE, message=FALSE}
stageR_res <- vector(mode="list",length = 5)
names(stageR_res) <- names(DD_res)

for (j in seq_along(DD_res)) {
  stageR_res_j <- vector(mode="list",length = 5)
  names(stageR_res_j) <- levels(pb_bin_ct$`B cell`$Status_on_day_collection_summary)[2:6]
  
  for(i in 1:5){
    pScreen <- rep(NA, nrow(DD_res[[j]][[i]]))
    for(h in 1:length(pScreen)){
        if(is.na(DD_res[[j]][[i]]$PValue[h])){
          pScreen[h] <- DGE_res[[j]][[i]]$PValue[h]
        } else {
          pScreen[h] <- hmp.stat(c(DD_res[[j]][[i]]$PValue[h],
                                   DGE_res[[j]][[i]]$PValue[h]),
                                 w=NULL)
        }
    }
    names(pScreen) <- rownames(DD_res[[j]][[i]])
    
    #confirmation stage
    pConfirmation <- as.matrix(cbind(DD_res[[j]][[i]]$PValue,
                                     DGE_res[[j]][[i]]$PValue))
    dimnames(pConfirmation) <- list(rownames(DD_res[[j]][[i]]), c("DD","DE"))
    
    # stageWise analysis
    stageRObj <- stageR(pScreen=pScreen, 
                        pConfirmation=pConfirmation, 
                        pScreenAdjusted=FALSE)
    stageRObj <- stageWiseAdjustment(object=stageRObj, 
                                     method="none", 
                                     alpha=0.05,
                                     allowNA = TRUE)
    res <- getResults(stageRObj)
    print(colSums(res)) #stage-wise analysis results
    
    if(colSums(res)[1] == 1){
      DD_DE_Sig_j_i <- getAdjustedPValues(stageRObj, 
                                      onlySignificantGenes=FALSE, 
                                      order=FALSE)
    } else{
      DD_DE_Sig_j_i <- getAdjustedPValues(stageRObj, 
                                      onlySignificantGenes=FALSE, 
                                      order=FALSE)
    }
    
    if(is.null(DD_DE_Sig_j_i)){
      DD_DE_Sig_j_i <- NA
    }
    
    stageR_res_j[[i]] <- DD_DE_Sig_j_i
  }
  stageR_res[[j]] <- stageR_res_j
}
```

## Manuscript table 1

With the following code, we generated *table1* from our manuscript.

```{r}
# number of differentially detected genes (BH-FDR, alpha < 0.05)
num_DD <- unlist(lapply(DD_res, function(ct){
    lapply(ct, function(x){
        sum(p.adjust(x$PValue, method="BH") < 0.05, na.rm = TRUE)
    })
}))

# number of differentially expressed genes (BH-FDR, alpha < 0.05)
num_DE <- unlist(lapply(DGE_res, function(ct){
    lapply(ct, function(x){
        sum(p.adjust(x$PValue, method="BH") < 0.05, na.rm = TRUE)
    })
}))

# number of overlap between differentially expressed and differentially detected genes 
num_overlap <- unlist(lapply(1:5, function(ct){
    lapply(1:5, function(contrast){
        sum(p.adjust(DD_res[[ct]][[contrast]]$PValue, method="BH") < 0.05 & 
                p.adjust(DGE_res[[ct]][[contrast]]$PValue, method="BH") < 0.05, na.rm=TRUE)
    })
}))

# number of genes that pass the omnibus test in the stageR analysis (BH-FDR, alpha < 0.05)
num_stagewise <- unlist(lapply(stageR_res, function(ct){
    lapply(ct, function(x){
        sum(x[,1] < 0.05, na.rm = TRUE)
    })
}))
table_1 <- data.frame(Comparison = 1:25,
                      "Cell type" = rep(c("B", "CsMB", "ImmB", "NaiveB", "UMB"), each = 5),
                      "COVID status" = rep(c("Asymptomatic", "Mild", "Moderate",
                                             "Severe", "Critical")),
                      DE = num_DE, 
                      DD = num_DD,
                      Overlap = num_overlap,
                      "Stage-wise" = num_stagewise,
                      row.names = NULL)
head(table_1)
```

## Manuscript supplementary table 2

With the following code, we generated *supplementary table 2* from our 
manuscript.

```{r}
# number of differentially detected genes in the second stage of the 
# stageR analysis (gene-level FDR, alpha < 0.05)
num_DD_stage2 <- unlist(lapply(stageR_res, function(ct){
    lapply(ct, function(x){
        sum(x[,2] < 0.05, na.rm = TRUE)
    })
}))

# overlap between DD genes from the regular analysis and DD genes from the
# stageR analysis
overlap_DD <- unlist(lapply(1:5, function(ct){
    lapply(1:5, function(contrast){
        sum(p.adjust(DD_res[[ct]][[contrast]]$PValue, method="BH") < 0.05 & 
                stageR_res[[ct]][[contrast]][,2] < 0.05, na.rm=TRUE)
    })
}))

# number of differentially expressed genes in the second stage of the 
# stageR analysis (gene-level FDR, alpha < 0.05)
num_DE_stage2 <- unlist(lapply(stageR_res, function(ct){
    lapply(ct, function(x){
        sum(x[,3] < 0.05, na.rm = TRUE)
    })
}))

# overlap between DE genes from the regular analysis and DE genes from the
# stageR analysis
overlap_DE <- unlist(lapply(1:5, function(ct){
    lapply(1:5, function(contrast){
        sum(p.adjust(DGE_res[[ct]][[contrast]]$PValue, method="BH") < 0.05 & 
                stageR_res[[ct]][[contrast]][,3] < 0.05, na.rm=TRUE)
    })
}))


table_sup2_v1 <- data.frame(Comparison = 1:25,
                      "Cell type" = rep(c("B", "CsMB", "ImmB", "NaiveB", "UMB"), each = 5),
                      "COVID status" = rep(c("Asymptomatic", "Mild", "Moderate",
                                             "Severe", "Critical")),
                      DE = num_DE,
                      DE_2 = num_DE_stage2,
                      overlap_DE_DE2 = overlap_DE,
                      DD = num_DD,
                      DD_2 = num_DD_stage2,
                      overlap_DD_DD2 = overlap_DD,
                      row.names = NULL)

table_sup2_v1[c(18,20,23,25),]
```

# Scatter plots and violin plots

## Naive Moderate

### Scatter plots

```{r}
scatterplotData_naive_mod <- scatterplot_data(data_bin = pb_bin_ct$`naive B cell`,
                                              data_count = pb_ct$`naive B cell`,
                                              DD_res = DD_res$`naive B cell`[[3]],
                                              DE_res = DGE_res$`naive B cell`[[3]],
                                              stageR_res = stageR_res$`naive B cell`$Moderate,
                                              group1 = "Healhty",
                                              group2 = "Moderate")

scatterplots_naive_mod <- scatterplots_visualize(plotData = scatterplotData_naive_mod,
                                                 limit_add = 1)

((scatterplots_naive_mod[[1]] + scatterplots_naive_mod[[2]]) / 
        (scatterplots_naive_mod[[3]] + scatterplots_naive_mod[[4]])) +
    plot_annotation(title = "Naive B cell - Healthy vs. moderately ill") + 
    plot_layout(guides = "collect")
```

### Violin plots

```{r, echo = FALSE, warning=FALSE, message=FALSE}
interest <- c("TRBC2", "PPIB", "ATP5F1A", "ACTG1")
violinplots_DD_naive_mod <- violin_DD(data_bin = pb_bin_ct$`naive B cell`,
                                      group1 = "Healthy",
                                      group2 = "Moderate",
                                      genes = interest,
                                      normalize = TRUE)
violinplots_DE_naive_mod <- violin_DGE(data_count = pb_ct$`naive B cell`,
                                       group1 = "Healthy",
                                       group2 = "Moderate",
                                       genes = interest)
```

### Manuscript figure

```{r}
set.seed(4634)
figures_naive_mod <- scatter_violin_paper(scatterplots = scatterplots_naive_mod,
                                          scatterplotData = scatterplotData_naive_mod,
                                          violinplots_DD = violinplots_DD_naive_mod,
                                          violinplots_DE = violinplots_DE_naive_mod,
                                          genes = c("TRBC2", "PPIB", "ATP5F1A"),
                                          legend_position = c(0.9,0.3),
                                          nudge_x = c(1,0,2),
                                          nudge_y = c(1,1,-1))

png("./figures/fig_covid_naiveModerate_complementarity.png",
            width     = 12,
            height    = 8,
            res = 600,
            units = "in")

((figures_naive_mod$topleft | figures_naive_mod$topright) + 
        plot_layout(guides = "keep")) /
    ((figures_naive_mod$bottomleft | figures_naive_mod$bottomright ) + 
         plot_layout(guides = "collect"))

dev.off()
```

## Naive Critical

### Scatter plots

```{r}
scatterplotData_naive_crit <- scatterplot_data(data_bin = pb_bin_ct$`naive B cell`,
                                               data_count = pb_ct$`naive B cell`,
                                               DD_res = DD_res$`naive B cell`[[5]],
                                               DE_res = DGE_res$`naive B cell`[[5]],
                                               stageR_res = stageR_res$`naive B cell`$Critical,
                                               group1 = "Healhty",
                                               group2 = "Critical")

scatterplots_naive_crit <- scatterplots_visualize(plotData = scatterplotData_naive_crit,
                                                  limit_add = 2)

((scatterplots_naive_crit[[1]] + scatterplots_naive_crit[[2]]) / 
        (scatterplots_naive_crit[[3]] + scatterplots_naive_crit[[4]])) +
    plot_annotation(title = "Naive B cell - Healthy vs. critically ill") + 
    plot_layout(guides = "collect")
```

### Violin plots

```{r, echo = FALSE, warning=FALSE, message=FALSE}
interest <- c("MKI67", "IGHV2-70D", "IFITM1")
violinplots_DD_naive_crit <- violin_DD(data_bin = pb_bin_ct$`naive B cell`,
                                       group1 = "Healthy",
                                       group2 = "Critical",
                                       genes = interest,
                                       normalize = TRUE)
violinplots_DE_naive_crit <- violin_DGE(data_count = pb_ct$`naive B cell`,
                                        group1 = "Healthy",
                                        group2 = "Critical",
                                        genes = interest)
```

### Manuscript figure

```{r}
set.seed(4634)
figures_naive_crit <- scatter_violin_paper(scatterplots = scatterplots_naive_crit,
                                           scatterplotData = scatterplotData_naive_crit,
                                           violinplots_DD = violinplots_DD_naive_crit,
                                           violinplots_DE = violinplots_DE_naive_crit,
                                           genes = c("MKI67", "IGHV2-70D", "IFITM1"),
                                           legend_position = c(0.9,0.73),
                                           nudge_x = c(1,0,0),
                                           nudge_y = c(0,2,-2))

png("./figures/fig_covid_naiveCritical_complementarity.png",
            width     = 12,
            height    = 8,
            res = 600,
            units = "in")

((figures_naive_crit$topleft | figures_naive_crit$topright) + 
        plot_layout(guides = "keep")) /
    ((figures_naive_crit$bottomleft | figures_naive_crit$bottomright ) + 
         plot_layout(guides = "collect"))

dev.off()
```

## Unswitched Moderate

### Scatter plots

```{r}
scatterplotData_unsw_mod <- scatterplot_data(data_bin = pb_bin_ct$`unswitched memory B cell`,
                                             data_count = pb_ct$`unswitched memory B cell`,
                                             DD_res = DD_res$`unswitched memory B cell`[[3]],
                                             DE_res = DGE_res$`unswitched memory B cell`[[3]],
                                             stageR_res = stageR_res$`unswitched memory B cell`$Moderate,
                                             group1 = "Healhty",
                                             group2 = "Moderate")

scatterplots_unsw_mod <- scatterplots_visualize(plotData = scatterplotData_unsw_mod,
                                                limit_add = 1)

((scatterplots_unsw_mod[[1]] + scatterplots_unsw_mod[[2]]) / 
        (scatterplots_unsw_mod[[3]] + scatterplots_unsw_mod[[4]])) +
    plot_annotation(title = "Unswitched memory B cell - Healthy vs. moderately ill") + 
    plot_layout(guides = "collect")
```

### Violin plots

```{r, echo = FALSE, warning=FALSE, message=FALSE}
interest <- c("S100A8","ATP8", "IGKV3-11")
violinplots_DD_unsw_mod <- violin_DD(data_bin = pb_bin_ct$`unswitched memory B cell`,
                                     group1 = "Healthy",
                                     group2 = "Moderate",
                                     genes = interest,
                                     normalize = TRUE)
violinplots_DE_unsw_mod <- violin_DGE(data_count = pb_ct$`unswitched memory B cell`,
                                      group1 = "Healthy",
                                      group2 = "Moderate",
                                      genes = interest)
```

### Manuscript figure

```{r}
set.seed(4634)
figures_unsw_mod <- scatter_violin_paper(scatterplots = scatterplots_unsw_mod,
                                         scatterplotData = scatterplotData_unsw_mod,
                                         violinplots_DD = violinplots_DD_unsw_mod,
                                         violinplots_DE = violinplots_DE_unsw_mod,
                                         genes = c("S100A8","ATP8", "IGKV3-11"),
                                         legend_position = c(0.9,0.3),
                                         nudge_x = c(1,1,-1),
                                         nudge_y = c(0,0,0))

png("./figures/fig_covid_unswModerate_complementarity.png",
            width     = 12,
            height    = 8,
            res = 600,
            units = "in")

((figures_unsw_mod$topleft | figures_unsw_mod$topright) + 
        plot_layout(guides = "keep")) /
    ((figures_unsw_mod$bottomleft | figures_unsw_mod$bottomright ) + 
         plot_layout(guides = "collect"))

dev.off()
```

## Unswitched Critical

### Scatter plots

```{r}
scatterplotData_unsw_crit <- scatterplot_data(data_bin = pb_bin_ct$`unswitched memory B cell`,
                                              data_count = pb_ct$`unswitched memory B cell`,
                                              DD_res = DD_res$`unswitched memory B cell`[[5]],
                                              DE_res = DGE_res$`unswitched memory B cell`[[5]],
                                              stageR_res = stageR_res$`unswitched memory B cell`$Critical,
                                              group1 = "Healhty",
                                              group2 = "Critical")

scatterplots_unsw_crit <- scatterplots_visualize(plotData = scatterplotData_unsw_crit,
                                                 limit_add = 1)

((scatterplots_unsw_crit[[1]] + scatterplots_unsw_crit[[2]]) / 
        (scatterplots_unsw_crit[[3]] + scatterplots_unsw_crit[[4]])) +
    plot_annotation(title = "Unswitched memory B cell - Healthy vs. critically ill") + 
    plot_layout(guides = "collect")
```

### Violin plots

We will not construct violin plots for this contrast, because there are no
differentially expressed genes

### Manuscript figure

```{r}
set.seed(4634)
figure_unsw_crit <- scatterplots_unsw_crit[[2]] +
    theme(strip.text.x = element_blank(),
      legend.position=c(0.92,0.2),
      legend.title = element_text(size = 10), 
      legend.text = element_text(size = 9),
      legend.spacing.y = unit(2, "mm"),
      legend.key.size = unit(0.5, 'lines')
      ) + 
  ggrepel::geom_label_repel( 
    data = scatterplotData_unsw_crit %>% filter(gene %in% c("EIF5A","HBB")),
    aes(label=gene),
    size = 3,
    fill = "white",
    color = "black",
    nudge_x = c(0,0),
    nudge_y = c(0,0),
    box.padding   = 1.5, 
    point.padding = 0.5,
    segment.color = 'grey50',
    min.segment.length = 0)  + 
  ggtitle("Panel A: pval-pval plot") + 
  theme(plot.title = element_text(size = 11, face="bold"))

png("./figures/fig_covid_unswCritical_complementarity.png",
            width     = 6,
            height    = 4,
            res = 600,
            units = "in")

figure_unsw_crit

dev.off()
```

# GSEA

```{r}
# load universe genes
universe <- openxlsx::read.xlsx("/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/allgenes.xlsx")$To

# Load link between symbols and Entrez
ensembl <- useEnsembl(biomart = 'genes',
                      dataset = 'hsapiens_gene_ensembl',
                      version = 109)
geneInfo <- getBM(attributes = c("hgnc_symbol",
                                 "entrezgene_id"),
                  mart = ensembl)
```

## Unswitched Moderate

### Stage 2 (manuscript table 2)

```{r}
# DD stage 2
gsea_DD2 <- run_gsea(genes = stageR_res$`unswitched memory B cell`$Moderate[,2],
                     geneInfo = geneInfo,
                     universe = universe)

# DE stage 2
gsea_DE2 <- run_gsea(genes = stageR_res$`unswitched memory B cell`$Moderate[,3],
                     geneInfo = geneInfo,
                     universe = universe)

# Omnibus
gsea_omnibus <- run_gsea(genes = stageR_res$`unswitched memory B cell`$Moderate[,1],
                         geneInfo = geneInfo,
                         universe = universe)

# Combine results from stageR omnibus test, stage 2 DD and stage 2 DE analysis
gsea_df_1 <- data.frame(Term = union(gsea_DD2$Term[1:10],
                                     union(gsea_DE2$Term[1:10],
                                           gsea_omnibus$Term[1:10])))
gsea_df_1$rank_DD2 <- gsea_DD2$Rank[match(gsea_df_1$Term, gsea_DD2$Term)]
gsea_df_1$rank_DE2 <- gsea_DE2$Rank[match(gsea_df_1$Term, gsea_DE2$Term)]
gsea_df_1$rank_omnibus <- gsea_omnibus$Rank[match(gsea_df_1$Term, gsea_omnibus$Term)]
gsea_df_1$pval_DD2 <- gsea_DD2$P.DE[match(gsea_df_1$Term, gsea_DD2$Term)]
gsea_df_1$pval_DE2 <- gsea_DE2$P.DE[match(gsea_df_1$Term, gsea_DE2$Term)]
gsea_df_1$pval_omnibus <- gsea_omnibus$P.DE[match(gsea_df_1$Term, gsea_omnibus$Term)]

# Write results file
# This file is used as table 2 in our manuscript
write.xlsx(x = gsea_df_1,
           file = "/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/Unsw_mod_stage2.xlsx")
```

### Stage 1 vs 2 (manuscript suppelmentary table 6)

```{r}
# DD
genes_DD <- p.adjust(DD_res$`unswitched memory B cell`[[3]]$PValue, method="BH")
names(genes_DD) <- rownames(DD_res$`unswitched memory B cell`[[3]])
gsea_DD <- run_gsea(genes = genes_DD,
                    geneInfo = geneInfo,
                    universe = universe)

# DE
genes_DE <- p.adjust(DGE_res$`unswitched memory B cell`[[3]]$PValue, method="BH")
names(genes_DE) <- rownames(DGE_res$`unswitched memory B cell`[[3]])
gsea_DE <- run_gsea(genes = genes_DE,
                    geneInfo = geneInfo,
                    universe = universe)

gsea_df_2 <- data.frame(Term = union(gsea_DD$Term[1:10],
                                     union(gsea_DD2$Term[1:10],
                                           union(gsea_DE$Term[1:10],
                                                 gsea_DE2$Term[1:10]))))
gsea_df_2$rank_DD <- gsea_DD$Rank[match(gsea_df_2$Term, gsea_DD$Term)]
gsea_df_2$rank_DD2 <- gsea_DD2$Rank[match(gsea_df_2$Term, gsea_DD2$Term)]
gsea_df_2$rank_DE <- gsea_DE$Rank[match(gsea_df_2$Term, gsea_DE$Term)]
gsea_df_2$rank_DE2 <- gsea_DE2$Rank[match(gsea_df_2$Term, gsea_DE2$Term)]
gsea_df_2$pval_DD <- gsea_DD$P.DE[match(gsea_df_2$Term, gsea_DD$Term)]
gsea_df_2$pval_DD2 <- gsea_DD2$P.DE[match(gsea_df_2$Term, gsea_DD2$Term)]
gsea_df_2$pval_DE <- gsea_DE$P.DE[match(gsea_df_2$Term, gsea_DE$Term)]
gsea_df_2$pval_DE2 <- gsea_DE2$P.DE[match(gsea_df_2$Term, gsea_DE2$Term)]

# Write results file
# This file is used as supplementary table 6 in our manuscript
write.xlsx(x = gsea_df_2,
           file = "/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/Unsw_mod_stage1vs2.xlsx")
```

## Unswitched Critical

### Stage 2 (manuscript supplementary table 3)

```{r}
# DD stage 2
gsea_DD2 <- run_gsea(genes = stageR_res$`unswitched memory B cell`$Critical[,2],
                     geneInfo = geneInfo,
                     universe = universe)

# DE and stagewise: can't be performed, no DE genes

gsea_df_3 <- data.frame(Term = gsea_DD2$Term,
                        rank_DD = gsea_DD2$Rank,
                        pval_DD = gsea_DD2$P.DE)

# Write results file
# This file is used as supplementary table 3 in our manuscript
write.xlsx(x = gsea_df_2[1:10,],
           file = "/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/Unsw_crit_stage2.xlsx")
```

### Stage 1 vs 2 (manuscript supplementary table 7)

```{r}
# DD
genes_DD <- p.adjust(DD_res$`unswitched memory B cell`[[5]]$PValue, method="BH")
names(genes_DD) <- rownames(DD_res$`unswitched memory B cell`[[5]])
gsea_DD <- run_gsea(genes = genes_DD,
                    geneInfo = geneInfo,
                    universe = universe)

gsea_df_4 <- data.frame(Term = union(gsea_DD$Term[1:10],
                                    gsea_DD2$Term[1:10]))
gsea_df_4$rank_DD <- gsea_DD$Rank[match(gsea_df_4$Term, gsea_DD$Term)]
gsea_df_4$rank_DD2 <- gsea_DD2$Rank[match(gsea_df_4$Term, gsea_DD2$Term)]

gsea_df_4$pval_DD <- gsea_DD$P.DE[match(gsea_df_4$Term, gsea_DD$Term)]
gsea_df_4$pval_DD2 <- gsea_DD2$P.DE[match(gsea_df_4$Term, gsea_DD2$Term)]

# Write results file
# This file is used as supplementary table 7 in our manuscript
write.xlsx(x = gsea_df_4,
           file = "/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/Unsw_crit_stage1vs2.xlsx")
```

## Naive Moderate

### Stage 2 (manuscript supplementary table 4)

```{r}
# DD stage 2
gsea_DD2 <- run_gsea(genes = stageR_res$`naive B cell`$Moderate[,2],
                     geneInfo = geneInfo,
                     universe = universe)

# DE stage 2
gsea_DE2 <- run_gsea(genes = stageR_res$`naive B cell`$Moderate[,3],
                     geneInfo = geneInfo,
                     universe = universe)

# Omnibus
gsea_omnibus <- run_gsea(genes = stageR_res$`naive B cell`$Moderate[,1],
                         geneInfo = geneInfo,
                         universe = universe)

gsea_df_5 <- data.frame(Term = union(gsea_DD2$Term[1:10], 
                                     union(gsea_DE2$Term[1:10],
                                           gsea_omnibus$Term[1:10])))
gsea_df_5$rank_DD2 <- gsea_DD2$Rank[match(gsea_df_5$Term, gsea_DD2$Term)]
gsea_df_5$rank_DE2 <- gsea_DE2$Rank[match(gsea_df_5$Term, gsea_DE2$Term)]
gsea_df_5$rank_omnibus <- gsea_omnibus$Rank[match(gsea_df_5$Term, gsea_omnibus$Term)]

gsea_df_5$pval_DD2 <- gsea_DD2$P.DE[match(gsea_df_5$Term, gsea_DD2$Term)]
gsea_df_5$pval_DE2 <- gsea_DE2$P.DE[match(gsea_df_5$Term, gsea_DE2$Term)]
gsea_df_5$pval_omnibus <- gsea_omnibus$P.DE[match(gsea_df_5$Term, gsea_omnibus$Term)]

# Write results file
# This file is used as supplementary table 4 in our manuscript
write.xlsx(x = gsea_df_5,
           file = "/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/naive_mod_stage2.xlsx")
```

### Stage 1vs2 (manuscript supplementary table 8)

```{r}
# DD
genes_DD <- p.adjust(DD_res$`naive B cell`[[3]]$PValue, method="BH")
names(genes_DD) <- rownames(DD_res$`naive B cell`[[3]])
gsea_DD <- run_gsea(genes = genes_DD,
                    geneInfo = geneInfo,
                    universe = universe)

# DE
genes_DE <- p.adjust(DGE_res$`naive B cell`[[3]]$PValue, method="BH")
names(genes_DE) <- rownames(DGE_res$`naive B cell`[[3]])
gsea_DE <- run_gsea(genes = genes_DE,
                    geneInfo = geneInfo,
                    universe = universe)

gsea_df_6 <- data.frame(Term = union(gsea_DD$Term[1:10], 
                                   union(gsea_DD2$Term[1:10],
                                         union(gsea_DE$Term[1:10],
                                               gsea_DE2$Term[1:10]))))
gsea_df_6$rank_DD <- gsea_DD$Rank[match(gsea_df_6$Term, gsea_DD$Term)]
gsea_df_6$rank_DD2 <- gsea_DD2$Rank[match(gsea_df_6$Term, gsea_DD2$Term)]
gsea_df_6$rank_DE <- gsea_DE$Rank[match(gsea_df_6$Term, gsea_DE$Term)]
gsea_df_6$rank_DE2 <- gsea_DE2$Rank[match(gsea_df_6$Term, gsea_DE2$Term)]
gsea_df_6$pval_DD <- gsea_DD$P.DE[match(gsea_df_6$Term, gsea_DD$Term)]
gsea_df_6$pval_DD2 <- gsea_DD2$P.DE[match(gsea_df_6$Term, gsea_DD2$Term)]
gsea_df_6$pval_DE <- gsea_DE$P.DE[match(gsea_df_6$Term, gsea_DE$Term)]
gsea_df_6$pval_DE2 <- gsea_DE2$P.DE[match(gsea_df_6$Term, gsea_DE2$Term)]

# Write results file
# This file is used as supplementary table 8 in our manuscript
write.xlsx(x = gsea_df_6,
           file = "/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/naive_mod_stage1vs2.xlsx")
```

## Naive Critical (manuscript supplementary table 5)

### Stage 2

```{r}
# DD stage 2
gsea_DD2 <- run_gsea(genes = stageR_res$`naive B cell`$Critical[,2],
                     geneInfo = geneInfo,
                     universe = universe)

# DE stage 2
gsea_DE2 <- run_gsea(genes = stageR_res$`naive B cell`$Critical[,3],
                     geneInfo = geneInfo,
                     universe = universe)

# stagewise
gsea_omnibus <- run_gsea(genes = stageR_res$`naive B cell`$Critical[,1],
                         geneInfo = geneInfo,
                         universe = universe)

gsea_df_7 <- data.frame(Term = union(gsea_DD2$Term[1:10], 
                                     union(gsea_DE2$Term[1:10],
                                           gsea_omnibus$Term[1:10])))

gsea_df_7$rank_DD <- gsea_DD2$Rank[match(gsea_df_7$Term, gsea_DD2$Term)]
gsea_df_7$rank_DE <- gsea_DE2$Rank[match(gsea_df_7$Term, gsea_DE2$Term)]
gsea_df_7$rank_omnibus <- gsea_omnibus$Rank[match(gsea_df_7$Term, gsea_omnibus$Term)]

gsea_df_7$pval_DD <- gsea_DD2$P.DE[match(gsea_df_7$Term, gsea_DD2$Term)]
gsea_df_7$pval_DE <- gsea_DE2$P.DE[match(gsea_df_7$Term, gsea_DE2$Term)]
gsea_df_7$pval_omnibus <- gsea_omnibus$P.DE[match(gsea_df_7$Term, gsea_omnibus$Term)]

# Write results file
# This file is used as supplementary table 5 in our manuscript
write.xlsx(x = gsea_df_7,
           file = "/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/naive_crit_stage2.xlsx")
```

### Stage 1 vs 2 (manuscript supplementary table 9)

```{r}
# DD
genes_DD <- p.adjust(DD_res$`naive B cell`[[5]]$PValue, method="BH")
names(genes_DD) <- rownames(DD_res$`naive B cell`[[5]])
gsea_DD <- run_gsea(genes = genes_DD,
                    geneInfo = geneInfo,
                    universe = universe)

# DE
genes_DE <- p.adjust(DGE_res$`naive B cell`[[5]]$PValue, method="BH")
names(genes_DE) <- rownames(DGE_res$`naive B cell`[[5]])
gsea_DE <- run_gsea(genes = genes_DE,
                    geneInfo = geneInfo,
                    universe = universe)

gsea_df_8 <- data.frame(Term = union(gsea_DD$Term[1:10], 
                                     union(gsea_DD2$Term[1:10],
                                           union(gsea_DE$Term[1:10],
                                                 gsea_DE2$Term[1:10]))))
gsea_df_8$rank_DD <- gsea_DD$Rank[match(gsea_df_8$Term, gsea_DD$Term)]
gsea_df_8$rank_DD2 <- gsea_DD2$Rank[match(gsea_df_8$Term, gsea_DD2$Term)]
gsea_df_8$rank_DE <- gsea_DE$Rank[match(gsea_df_8$Term, gsea_DE$Term)]
gsea_df_8$rank_DE2 <- gsea_DE2$Rank[match(gsea_df_8$Term, gsea_DE2$Term)]
gsea_df_8$pval_DD <- gsea_DD$P.DE[match(gsea_df_8$Term, gsea_DD$Term)]
gsea_df_8$pval_DD2 <- gsea_DD2$P.DE[match(gsea_df_8$Term, gsea_DD2$Term)]
gsea_df_8$pval_DE <- gsea_DE$P.DE[match(gsea_df_8$Term, gsea_DE$Term)]
gsea_df_8$pval_DE2 <- gsea_DE2$P.DE[match(gsea_df_8$Term, gsea_DE2$Term)]

# Write results file
# This file is used as supplementary table 9 in our manuscript
write.xlsx(x = gsea_df_8,
           file = "/Users/jg/Desktop/PhD/DD_project/DD_cases/Covid/GSEA/naive_crit_stage1vs2.xlsx")
```

# Visualize offset

This figure displays the normalization offsets (CDR) that were used in the 
differential detection inference. Note that this figure was not included in the
manuscript.

```{r}
pdf("./figures/casecovid_CDR.pdf",
            width     = 10,
            height    = 7,
            pointsize = 3)

titles <- names(pb_bin_ct)
plotList <- vector(mode = "list", length = 5)
for (i in seq_along(pb_bin_ct)) {
    element <- pb_bin_ct[[i]]
    bin_counts <- assay(element)
    of <- colMeans(sweep(bin_counts, 2, element$ncells, "/")) 
    logitOf <- log(of/(1-of))
    df <- data.frame(status = element$Status_on_day_collection_summary,
                     offset = of,
                     Site = element$Site)
    
    plotList[[i]] <- ggplot(data = df, aes(x = status, y = offset, col = Site)) +
        geom_jitter(width = 0.2) +
        theme_bw() +
        ggtitle(titles[i]) + 
        theme(plot.title = element_text(size = 12),
              axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
}

(plotList[[1]] + plotList[[2]] + plotList[[3]]) / 
    (plotList[[4]] + plotList[[5]] + plot_spacer()) + plot_layout(guides = "collect")

dev.off()
```

# Get percentage 0-1-2-2+ for single-cell level data

Here we obtain the fraction of counts that are zeros, ones, twos and values 
higher than two in the single-cell level data. This information is used in the
discussion section of our manuscript.

```{r}
pb_ct <- readRDS("./objects/pb_ct_filt.rds")
sce <- readRDS("./objects/sce_Covid_Bcells.rds")
```

```{r}
# B cell
sce_hlp <- sce[,sce$cell_type_curated == "B cell"]
sce_hlp <- sce_hlp[rownames(pb_ct$`B cell`),]
sce_hlp <- sce_hlp[,sce_hlp$donor_id %in% pb_ct$`B cell`$donor_id]
a_entries <- nrow(sce_hlp)*ncol(sce_hlp)
a_0 <- sum(assay(sce_hlp) == 0)
a_1 <- sum(assay(sce_hlp) == 1)
a_2 <- sum(assay(sce_hlp) == 2)
a_r <- sum(assay(sce_hlp) > 2)

# class switched memory B cell
sce_hlp <- sce[,sce$cell_type_curated == "class switched memory B cell"]
sce_hlp <- sce_hlp[rownames(pb_ct$`class switched memory B cell`),]
sce_hlp <- sce_hlp[,sce_hlp$donor_id %in% pb_ct$`class switched memory B cell`$donor_id]
b_entries <- nrow(sce_hlp)*ncol(sce_hlp)
b_0 <- sum(assay(sce_hlp) == 0)
b_1 <- sum(assay(sce_hlp) == 1)
b_2 <- sum(assay(sce_hlp) == 2)
b_r <- sum(assay(sce_hlp) > 2)

# immature B cell
sce_hlp <- sce[,sce$cell_type_curated == "immature B cell"]
sce_hlp <- sce_hlp[rownames(pb_ct$`immature B cell`),]
sce_hlp <- sce_hlp[,sce_hlp$donor_id %in% pb_ct$`immature B cell`$donor_id]
c_entries <- nrow(sce_hlp)*ncol(sce_hlp)
c_0 <- sum(assay(sce_hlp) == 0)
c_1 <- sum(assay(sce_hlp) == 1)
c_2 <- sum(assay(sce_hlp) == 2)
c_r <- sum(assay(sce_hlp) > 2)

# naive B cell
sce_hlp <- sce[,sce$cell_type_curated == "naive B cell"]
sce_hlp <- sce_hlp[rownames(pb_ct$`naive B cell`),]
sce_hlp <- sce_hlp[,sce_hlp$donor_id %in% pb_ct$`naive B cell`$donor_id]
d_entries <- nrow(sce_hlp)*ncol(sce_hlp)
d_0 <- sum(assay(sce_hlp) == 0)
d_1 <- sum(assay(sce_hlp) == 1)
d_2 <- sum(assay(sce_hlp) == 2)
d_r <- sum(assay(sce_hlp) > 2)

# unswitched memory B cell
sce_hlp <- sce[,sce$cell_type_curated == "unswitched memory B cell"]
sce_hlp <- sce_hlp[rownames(pb_ct$`unswitched memory B cell`),]
sce_hlp <- sce_hlp[,sce_hlp$donor_id %in% pb_ct$`unswitched memory B cell`$donor_id]
e_entries <- nrow(sce_hlp)*ncol(sce_hlp)
e_0 <- sum(assay(sce_hlp) == 0)
e_1 <- sum(assay(sce_hlp) == 1)
e_2 <- sum(assay(sce_hlp) == 2)
e_r <- sum(assay(sce_hlp) > 2)
```

```{r}
(a_0+b_0+c_0+d_0+e_0)/(a_entries+b_entries+c_entries+d_entries+e_entries)*100
(a_1+b_1+c_1+d_1+e_1)/(a_entries+b_entries+c_entries+d_entries+e_entries)*100
(a_2+b_2+c_2+d_2+e_2)/(a_entries+b_entries+c_entries+d_entries+e_entries)*100
(a_r+b_r+c_r+d_r+e_r)/(a_entries+b_entries+c_entries+d_entries+e_entries)*100
```

# Session info

```{r}
sessionInfo()
```


